source/guides/recipes/dependency-management.md

```diff
@@ -0,0 +1,101 @@
+(dependency-management-niv)=
+# Managing remote sources with niv
+
+The Nix language can be used to describe dependencies between files managed by Nix.
+Nix expressions themselves can depend on remote sources, and there are multiple ways to specify their origin, as shown in [](pinning-nixpkgs).
+
+For more automation around handling remote sources, set up [niv](https://github.com/nmattia/niv/) in your project:
+
+```shell-session
+$ nix-shell -p niv --run "niv init"
```

Comment:
Oops. This is the weirdest thing. Running that doesn't actually work! (To make sure, here I also specified the source for `niv`):
```shellSession
$ nix-shell -I nixpkgs=channel:nixos-23.05 -p niv --run "niv init --nixpkgs nixos/nixpkgs --nixpkgs-branch nixos-23.05"
Initializing
  Creating nix/sources.nix
  Creating nix/sources.json
  Importing 'nixpkgs' ...
  Adding package nixpkgs
    FATAL: Could not read the output of 'nix-prefetch-url'. This is a bug. Please create a
    ticket:
    
      https://github.com/nmattia/niv/issues/new
    
    Thanks! I'll buy you a beer.
    command: 
    nix-prefetch-url--unpack https://github.com/nixos/nixpkgs/archive/34bdaaf1f0b7fb6d9091472edc968ff10a8c2857.tar.gz --name 34bdaaf1f0b7fb6d9091472edc968ff10a8c2857.tar.gz
    stdout: 
    
    stderr: 
    error: failed to extract archive (Write failed)
    
    FATAL: One or more packages failed to update:
    nixpkgs: ExitFailure 1
  ERROR: ExitFailure 1
ERROR: ExitFailure 1
```

The command sample that is output *looks* broken, but it's just [a little representation bug here](https://github.com/nmattia/niv/blob/723f0eeb969a730db3c30f977c2b66b9dce9fe4a/src/Niv/GitHub/Cmd.hs#L187):

```shellSession
$ nix-prefetch-url --unpack https://github.com/nixos/nixpkgs/archive/34bdaaf1f0b7fb6d9091472edc968ff10a8c2857.tar.gz --name 34bdaaf1f0b7fb6d9091472edc968ff10a8c2857.tar.gz
```
Running this in the shell from which we originally called `nix-shell` will in fact succeed. But that's a different environment. If we try the pure shell, and drop into it [during the `--command`](https://nixos.org/manual/nix/stable/command-ref/nix-shell#options), to attempt to run that command, it will fail (don't forget to clean the `nix` directory left over from the previous call):

 ```shellSession
$ rm -rf nix
$ nix-shell -I nixpkgs=channel:nixos-23.05 -p niv nix cacert --pure --command "niv init --nixpkgs nixos/nixpkgs --nixpkgs-branch nixos-23.05; return"
# ... same error as above
[nix-shell]$ nix-prefetch-url --unpack https://github.com/nixos/nixpkgs/archive/34bdaaf1f0b7fb6d9091472edc968ff10a8c2857.tar.gz --name 34bdaaf1f0b7fb6d9091472edc968ff10a8c2857.tar.gz
error: failed to extract archive (Write failed)
```

Without looking into the C++ I'd guess that something about the environment in `nix-shell` is different that prevents `nixpkgs-prefetch-url` to write to the store. Even weirder is that this is only true for `NixOS/nixpkgs`, but not for a smaller repo such as used in https://github.com/NixOS/nix/issues/5614.

Proof. For a fair comparison, init with empty sources and then add `nixpkgs` (fails), then start over and add `haskell.nix`:

```shellSession
$ rm -rf nix
$ nix-shell -I nixpkgs=channel:nixos-23.05 -p niv nix cacert --pure --run "niv init --no-nixpkgs; niv add nixos/nixpkgs -b nixos-23.05"
# ... same error
```

```shellSession
$ rm -rf nix
$ nix-shell -I nixpkgs=channel:nixos-23.05 -p niv nix cacert --pure --run "niv init --no-nixpkgs; niv add nixos/nixpkgs -b nixos-23.05"
# ... same error
$ rm -rf nix
$ nix-shell -I nixpkgs=channel:nixos-23.05 -p niv nix cacert --pure --run "niv init --no-nixpkgs; niv add input-output-hk/haskell.nix -r aaf4e230c69507fb3a1c6fe4c1f6603d659d5543"
# ...
Done: Adding package haskell.nix
```

One must assume there is something in `nix-prefetch-url` that will behave differently with a large download in a `nix-shell` environment.

Reply:
Oh wow, I indeed ran into `/tmp` being full!

Thanks to these two helpful comments this was not entirely impossible to debug:
- https://github.com/DavHau/mach-nix/issues/458#issuecomment-1248583956
- https://github.com/NixOS/nix/issues/5857#issuecomment-1509862011

The funniest part is the reason how it even got there: When we tested the kernel build instructions with @StefanSchroeder earlier, the kernel *source code* filled up my `tmpfs` with 1.5GiB. Now that's a massive codebase.

Reply:
Damn, nice find! I guess in the end it's unrelated and we shouldn't worry about it :)